### ISSUE TITLE
Multiple executions of Vue appEntrypoint. One per component.

### ISSUE THREAD

**Issue by angelcervera** [withastro/astro#6312]
### What version of `astro` are you using?

2.0.4

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### What operating system are you using?

Linux

### Describe the Bug

In the Vue integration, `appEntrypoint` is executed one time per component, so it is not possible to use the officially recommended [Pinia](https://pinia.vuejs.org/) because looks Astro is creating one isolated storage per component.

In case this is the expected behaviour in Astro Vue integration, where to run the `createPinia()` to be able to create the shared storage properly?

### Link to Minimal Reproducible Example

https://stackblitz.com/github/angelcervera/astro_plus_vue_init_bug

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by felix-berlin** [comment]
As alternativ you can try [nano stores](https://docs.astro.build/en/core-concepts/sharing-state/#why-nano-stores).

**Comment by angelcervera** [comment]
@felix-berlin Thanks for your advice. In any case, if this is a bug, it is a huge problem, not only because Pinia.

**Comment by angelcervera** [comment]
And wait. Astro is going to run the appEntrypoint even when the plugin is not used in the page at all!!!! I think that I'm missing something.

**Comment by louiss0** [comment]
> And wait. Astro is going to run the `appEntrypoint` even when the plugin is not used in the page at all!!!! I think that I'm missing something.

Yes it's because that is how island architecture works. An new instance is created per component. By the way, you can use `useSharedComposable()`. I never thought about this that much but I think you are right I don't think there is a way around this.

**Comment by angelcervera** [comment]
Hi @louiss0 What is useSharedComposable()? Another alternative like nanostores? Could you share any reference or doc? It you are talking about [Vue Use State Effect](https://github.com/lukasborawski/vue-use-state-effect), I will have the same problem because is sharing states between components inside the same app, and Astro is creating a new app per lets call it `root` component.

**Comment by louiss0** [comment]
> Hi @louiss0 What is useSharedComposable()? Another alternative like nanostores? Could you share any reference or doc? If you are talking about [Vue Use State Effect](https://github.com/lukasborawski/vue-use-state-effect), I will have the same problem because is sharing states between components inside the same app, and Astro is creating a new app per lets call it `root` component.


Vue Use Shared Composable should solve that problem for you Here's the docs https://vueuse.org/shared/createSharedComposable/#createsharedcomposable.

**Comment by angelcervera** [comment]
WooW!! The problem is even worse. The `appEntrypoint` script is executed one time per component and per page when building, no matter if you are using Vue only in the client.
I'm trying Astro with a site with more than 100K pages, so when building, the script is executed more than 300K times. 😱

**Comment by angelcervera** [comment]
Hi @natemoo-re Sorry for mentioning you directly, but I saw you as the main committer in this part of Astro. So maybe you are interested in this.

To resume:
1. When root components are `client:*`,  `appEntrypoint` is executed both locally when building and in the client. I would expect running only in the client. It means: the script is executed one time per page. If you have 100K pages, it is going to run 100K times even if it is never used. There are Vue plugins like firebase related or vue-instantsearch, that are not going to be used in the server when building the static part, and they are taking a lot of time non-necessary extra time (which means hours of extra building time).
2. The `appEntrypoint` script is running one time per root component, no matter if the component will use these plugins or not because the script is common to all Apps instantiated (It is, one per root component). It means, multiple inits or even instances of the same plugin on the same page.
3. If Astro creates one Vue App per Island, it should provide a way to specify the `appEntrypoint` per Island (root component).

I don't know if this is a bug. But it is not, maybe we can improve the behavior.

**Comment by matthewp** [comment]
It's common to use Vue plugins on the server too. If you want them to only run in the client you can use `import.meta.env.SSR` to prevent adding those in the server.

**Comment by matthewp** [comment]
Closing as this is expected.

**Comment by angelcervera** [comment]
Hi @matthewp I think that you did not get the point at all. It is not about running it in the server or in the client. It is about sharing the same init for all islands.

If every island is going to create a Vue App and init the same script, you can not isolate every execution when it is the idea of an island.

Make not sense to have different Vue instances running all the same init.

I found the issue because the same init script was running multiple times in the client (not only in the server), but the issue is that the `appEntrypoint` is shared for the all Islands.

Imagine that you have a Vue plugin running on one page and other on another page. You are not able to init one but not the other, because there is only one app `appEntrypoint`.
For example, I need to init the Algolia plugin, but not on all pages. But even if I want to use my own Vue components, Algolia is initializing. Do you get my point? It is nothing related to running or not in the server.

That is the core of the problem. And it is so weird that other did not realize it that I'm thinking maybe I'm missing something.

**Comment by adopted3** [comment]
Why is this closed? It's important.

**Comment by AngeloFornerino** [comment]
Hi @matthewp, has there been any updates to this issue?

**Comment by kissu** [comment]
I also thought that if using `transition:persist` + `transition:persist-props` + `client:load`, it would persist the state of Pinia without resetting it, but I guess it doesn't. Resets to the initial value because a new Pinia instance is still created. 🤔 
https://docs.astro.build/en/guides/view-transitions/#transitionpersist-props
